# Extra cornetto steps for saliva samples

## Removing non-human reads from saliva samples

For human saliva samples, make sure you have removed non-human reads before running hifiasm. We used centrifuge for this:
```bash
# run centrifuge
centrifuge -p 48 -q -x index/p_compressed+h+v -U reads-0_all.fastq -S reads_classification.tsv --report-file reads_report.tsv

# extract human reads
samtools faidx reads-0_all.fastq
cat reads_classification.tsv | awk '$3!=9606' | cut -f 1 | sort -u > nonhuman_reads.txt
cut -f 1 reads-0_all.fai > all_reads.txt
grep -v -F -f nonhuman_reads.txt all_reads.txt > human_reads.txt
samtools fqidx -r human_reads.txt reads-0_all.fastq > reads-0.fastq
```

## Get the non-human contigs

For human saliva samples, we need to append the non human contigs to each cornetto iteration. For this, follow the steps below:

1. Identify nonhuman species with minimum 100 reads, based on `reads_report.tsv` we generated [above](#removing-non-human-reads-from-saliva-samples).

```bash
cat reads_report.tsv | sed 's/ /-/g' | sort -k5,5nr | awk '$2 != 9606' | awk '$5 >= 100' | cut -f 2 | sort -u | awk '$1 != "taxID"' > nonhuman_species_high_count.txt
```

2. Now assemble all the reads in the base assembly (both human and non human) with hifiasm.

3. Run centrifuge on the assembly generated by step 2 above.

```bash
centrifuge -p 48 -f -x index_dir/p_compressed+h+v -U asm-all-0.fasta -S ${CLASSIFICATION} --report-file ${REPORT}
```

3. Find the contig IDs from in the assembly from step 2 above (`asm-all-0.fasta`), that correspond to nonhuman species detected in step 1. Extract these contigs into a fasta file called `asm-all-0.nonhuman_contigs.fasta`, with the contig names appended with "_nonhuman".


4. create a BED file covering full contigs for each nonhuman contig to be excluded during readfish
```bash
samtools faidx asm-all-0.nonhuman_contigs.fasta
awk '{print $1"\t0\t"$2}' asm-0.fasta.fai > asm-all-0.nonhuman_contigs.bed
```